### PR TITLE
Search additional roots

### PR DESCRIPTION
**Pull request by gbodeen**
Signed-off-by: Gabriel Bodeen <gabriel.bodeen@ericsson.com>

#### What it does

This adds a preference for additional root directories to search in both the Search-In-Workspace widget and the File-Search (ctrl+p) menu.

In interviews with our users, one of their requests was for the search features to be able to find code that wasn't technically within their workspace, but that was located in other, closely-related directories. For particularly large directories, it can be advantageous to make the code easily searchable with Theia without incurring other costs such as file watching.

I wanted to check whether there was appetite for this preference to be added directly to the Theia project.  This preference explicitly goes beyond what is currently available in VS Code.

#### How to test

1. Build Theia and launch with the Theia repository as the workspace.
2. Search the preferences for `additional`, switch to Workspace scope, and edit the JSON to enter one or more paths, e.g. to the Blueprint repo and to a non-existent directory. You can use variable substitutions. For example, in my settings.json I added `"search.additionalRoots": ["${env:HOME}/test/", "/repo/${env:USER}/theia-blueprint"]`.
3. Use the Search-In-Workspace widget and the File-Search (ctrl+p) menu with various terms and filenames. You should observe entries from both the Theia and Blueprint repos; the presence of a non-existent directory should not cause any errors.

#### Review checklist

- [X] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

** Note: I haven't tested on Windows or Mac.

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/search-in-workspace/src/browser/search-in-workspace-preferences.ts [hunk 1]
```diff
@@ -58,6 +58,11 @@ export const searchInWorkspacePreferencesSchema: PreferenceSchema = {
             description: nls.localizeByDefault('Controls whether to follow symlinks while searching.'),
             default: true,
             type: 'boolean',
+        },
+        'search.additionalRoots': {
+            description: nls.localizeByDefault('Directories outside the current workspace that should be included in searches.'),
```
Comment (vince-fugnitto) [inline packages/search-in-workspace/src/browser/search-in-workspace-preferences.ts:63]: I do not believe there is a default value for this string in `nls.metada.json`:

```suggestion
            description: nls.localize('theia/search-in-workspace/additionalRoots' 'Directories outside the current workspace that should be included in searches.'),
```

File: packages/search-in-workspace/src/browser/search-in-workspace-service.ts [hunk 2]
```diff
@@ -72,6 +73,7 @@ export class SearchInWorkspaceService implements SearchInWorkspaceClient {
     @inject(SearchInWorkspaceClientImpl) protected readonly client: SearchInWorkspaceClientImpl;
     @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService;
     @inject(ILogger) protected readonly logger: ILogger;
+    @inject(SearchInWorkspacePreferences) protected readonly preferenceProxy: SearchInWorkspacePreferences;
```
Comment (vince-fugnitto) [inline packages/search-in-workspace/src/browser/search-in-workspace-service.ts:76]: minor:

```suggestion
    @inject(SearchInWorkspacePreferences) protected readonly searchInWorkspacePreferences: SearchInWorkspacePreferences;
```

**Review by vince-fugnitto — COMMENTED** [review]
@gbodeen speaking offline with @colin-grant-work, I believe in order for the pull-request to work well, we might need some variable resolution in `settings.json`. I'm not sure that setting absolute paths will work well in the `workspace` scope as this is usually checked-in to the repo, and will not work for other users.

**Comment by gbodeen** [comment]
> @gbodeen speaking offline with @colin-grant-work, I believe in order for the pull-request to work well, we might need some variable resolution in `settings.json`. I'm not sure that setting absolute paths will work well in the `workspace` scope as this is usually checked-in to the repo, and will not work for other users.

OK, I'll take a look soon.

**Comment by gbodeen** [comment]
@vince-fugnitto Variable resolution is added now.

**Review by alvsan09 — APPROVED** [review]
I have tested the following, and it works as expected :+1: 
* Search of files (ctrl+p) within the folder of the `additionalRoots`
* Search  in workspace results include files withing the configured `additionalRoots`

The code looks good to me as well, 
Thanks !!

**Comment by vince-fugnitto** [comment]
@gbodeen do you mind rebasing the pull-request, I can take a final look afterwards :)

**Comment by vince-fugnitto** [comment]
> I'm not sure whether there's a variation on that which would be more generally useful and so suitable for Theia.

@gbodeen sounds good! If you don't feel that the feature is useful in the base framework I think we can close the pr for the moment.

**Pull request closed without merging**